cpp/CMakeLists.txt

```diff
@@ -15,7 +15,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
-cmake_minimum_required(VERSION 3.5)
+cmake_minimum_required(VERSION 3.12)
```

Comment:
Sorry for confusing you.
We can't upgrade required CMake version to 3.12 uncoditionally. Because Ubuntu 18.04 still uses CMake 3.10.

If we use the object library approach, we report an error with CMake < 3.12 like https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L1396-L1399 .

But I think that the project global `CMAKE_CXX_STANDARD=14` approach is better than the object library approach for maintenance as I described at https://github.com/apache/arrow/pull/12914#discussion_r899724290 .

How about reverting `add_library(OBJECT)` related changes?